From ec0b866cd73c8200f1399c08114fbda08d3072d1 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Tue, 25 Mar 2025 11:38:08 +0100
Subject: [PATCH] [SpeechSynthesis] Emit error event for all cancelled
 utterances

Dispatch an error event "canceled" for queued utterances.

Spec:
https://dvcs.w3.org/hg/speech-api/raw-file/tip/webspeechapi
5.2.5.1 SpeechSynthesisErrorEvent Attributes:

"canceled"
    A cancel method call caused the SpeechSynthesisUtterance
    to be removed from the queue before it had begun being spoken.
---
 .../WebCore/Modules/speech/SpeechSynthesis.cpp   | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/Modules/speech/SpeechSynthesis.cpp b/Source/WebCore/Modules/speech/SpeechSynthesis.cpp
index 3ab9c98261f59..7a803219c096d 100644
--- a/Source/WebCore/Modules/speech/SpeechSynthesis.cpp
+++ b/Source/WebCore/Modules/speech/SpeechSynthesis.cpp
@@ -169,16 +169,28 @@ void SpeechSynthesis::cancel()
     // Remove all the items from the utterance queue.
     // Hold on to the current utterance so the platform synthesizer can have a chance to clean up.
     RefPtr current = protectedCurrentSpeechUtterance();
-    m_utteranceQueue.clear();
     if (m_speechSynthesisClient) {
+        m_utteranceQueue.clear();
         m_speechSynthesisClient->cancel();
         // If we wait for cancel to callback speakingErrorOccurred, then m_currentSpeechUtterance will be null
         // and the event won't be processed. Instead we process the error immediately.
         speakingErrorOccurred(SpeechSynthesisErrorCode::Canceled);
         m_currentSpeechUtterance = nullptr;
-    } else if (m_platformSpeechSynthesizer)
+    } else if (m_platformSpeechSynthesizer) {
+        // Clear m_utteraceQueue before calling cancel to avoid picking up new utterances
+        // on platform completition callback
+        Deque<Ref<SpeechSynthesisUtterance>> utteranceQueue = WTFMove(m_utteranceQueue);
         m_platformSpeechSynthesizer->cancel();
 
+        // Trigger canceled events for queued utterances
+        while (utteranceQueue.size() > 0) {
+            Ref<SpeechSynthesisUtterance> utterance = utteranceQueue.takeFirst();
+            // Current utterance is handled in platform cancel()
+            if (current.get() != utterance.ptr())
+                utterance.get().errorEventOccurred(eventNames().errorEvent, SpeechSynthesisErrorCode::Canceled);
+        }
+    }
+
     current = nullptr;
 }
 
